File: stress/stress/backend_stressng.py
"""StressBackend implementation using stress-ng for real SP-1/SP-2 injection."""
from __future__ import annotations

import subprocess
from typing import Optional

from stress.stress.backend import (
    StressBackend, NetworkDegradationConfig, ResourcePressureConfig, MemoryStressConfig,
)
from stress.stress.backend_linux import LinuxBackend


class StressNgBackend(StressBackend):
    """Real stress injection via stress-ng subprocess.

    Covers SP-1 (radiation/bit-flip via --vm-method flip) and
    SP-2 (thermal cycling via --cpu-load modulation).
    Delegates network, power, and isolation to LinuxBackend.
    """

    def __init__(self) -> None:
        self._linux = LinuxBackend()
        self._memory_proc: Optional[subprocess.Popen] = None
        self._cpu_proc: Optional[subprocess.Popen] = None

    def inject_memory_stress(self, config: MemoryStressConfig) -> None:
        self.remove_memory_stress()
        self._memory_proc = subprocess.Popen([
            "stress-ng",
            "--vm", "1",
            "--vm-bytes", str(config.vm_bytes),
            "--vm-method", config.vm_method,
            "--vm-keep",
        ], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

    def remove_memory_stress(self) -> None:
        if self._memory_proc is not None:
            self._memory_proc.terminate()
            self._memory_proc.wait()
            self._memory_proc = None

    def inject_cpu_stress(self, cpu_load: int) -> None:
        """SP-2 thermal cycling: set CPU load percentage via stress-ng."""
        self.remove_cpu_stress()
        self._cpu_proc = subprocess.Popen([
            "stress-ng",
            "--cpu", "0",
            "--cpu-load", str(cpu_load),
        ], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

    def remove_cpu_stress(self) -> None:
        if self._cpu_proc is not None:
            self._cpu_proc.terminate()
            self._cpu_proc.wait()
            self._cpu_proc = None

    def apply_network_degradation(self, config: NetworkDegradationConfig) -> None:
        self._linux.apply_network_degradation(config)

    def remove_network_degradation(self, interface: str) -> None:
        self._linux.remove_network_degradation(interface)

    def apply_resource_pressure(self, config: ResourcePressureConfig) -> None:
        self._linux.apply_resource_pressure(config)

    def remove_resource_pressure(self, cgroup_path: str) -> None:
        self._linux.remove_resource_pressure(cgroup_path)

    def pause_workload(self, pid: int) -> None:
        self._linux.pause_workload(pid)

    def resume_workload(self, pid: int) -> None:
        self._linux.resume_workload(pid)

    def apply_network_partition(self, interface: str) -> None:
        self._linux.apply_network_partition(interface)

    def remove_network_partition(self, interface: str) -> None:
        self._linux.remove_network_partition(interface)

    def name(self) -> str:
        return "stress-ng"

File: stress-ref/src/stress/backend_stressng.rs
use std::process::{Child, Command, Stdio};
use crate::stress::backend::*;

pub struct StressNgBackend {
    linux: crate::stress::backend_linux::LinuxBackend,
    memory_proc: Option<Child>,
    cpu_proc: Option<Child>,
}

impl StressNgBackend {
    pub fn new() -> Self {
        Self {
            linux: crate::stress::backend_linux::LinuxBackend,
            memory_proc: None,
            cpu_proc: None,
        }
    }

    pub fn inject_cpu_stress(&mut self, cpu_load: u32) -> Result<(), String> {
        self.remove_cpu_stress();
        let child = Command::new("stress-ng")
            .args(["--cpu", "0", "--cpu-load", &cpu_load.to_string()])
            .stdout(Stdio::null()).stderr(Stdio::null())
            .spawn().map_err(|e| format!("stress-ng spawn failed: {e}"))?;
        self.cpu_proc = Some(child);
        Ok(())
    }

    pub fn remove_cpu_stress(&mut self) {
        if let Some(mut p) = self.cpu_proc.take() {
            let _ = p.kill();
            let _ = p.wait();
        }
    }
}

impl StressBackend for StressNgBackend {
    fn inject_memory_stress(&mut self, config: &MemoryStressConfig) -> Result<(), String> {
        self.remove_memory_stress()?;
        let child = Command::new("stress-ng")
            .args([
                "--vm", "1",
                "--vm-bytes", &config.vm_bytes.to_string(),
                "--vm-method", &config.vm_method,
                "--vm-keep",
            ])
            .stdout(Stdio::null()).stderr(Stdio::null())
            .spawn().map_err(|e| format!("stress-ng spawn failed: {e}"))?;
        self.memory_proc = Some(child);
        Ok(())
    }

    fn remove_memory_stress(&mut self) -> Result<(), String> {
        if let Some(mut p) = self.memory_proc.take() {
            let _ = p.kill();
            let _ = p.wait();
        }
        Ok(())
    }

    fn pause_workload(&mut self, pid: u32) -> Result<(), String> {
        self.linux.pause_workload(pid)
    }

    fn resume_workload(&mut self, pid: u32) -> Result<(), String> {
        self.linux.resume_workload(pid)
    }

    fn apply_network_degradation(&mut self, config: &NetworkDegradationConfig) -> Result<(), String> {
        self.linux.apply_network_degradation(config)
    }

    fn remove_network_degradation(&mut self, interface: &str) -> Result<(), String> {
        self.linux.remove_network_degradation(interface)
    }

    fn apply_resource_pressure(&mut self, config: &ResourcePressureConfig) -> Result<(), String> {
        self.linux.apply_resource_pressure(config)
    }

    fn remove_resource_pressure(&mut self, cgroup_path: &str) -> Result<(), String> {
        self.linux.remove_resource_pressure(cgroup_path)
    }

    fn apply_network_partition(&mut self, interface: &str) -> Result<(), String> {
        self.linux.apply_network_partition(interface)
    }

    fn remove_network_partition(&mut self, interface: &str) -> Result<(), String> {
        self.linux.remove_network_partition(interface)
    }

    fn name(&self) -> &str { "stress-ng" }
}

impl Drop for StressNgBackend {
    fn drop(&mut self) {
        let _ = self.remove_memory_stress();
        self.remove_cpu_stress();
    }
}
